### PR TITLE
SAAS-2062: Plugin to check for outdated Amazon Machine Images

### PR DESCRIPTION
**Pull request by UmairAjmal**

### CODE REVIEW

File: plugins/aws/ec2/outdatedAmi.js [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        }
+    },
+
+    run: function(cache, settings, callback) {
+        var config = {
+            check_ami_usage: settings.check_ami_usage || this.settings.check_ami_usage.default
+        };
+
+        config.check_ami_usage = (config.check_ami_usage == 'true');
+
```
Comment (AkhtarAmir) [inline plugins/aws/ec2/outdatedAmi.js:27]: Check if (!config.check_ami_usage) return callback(null, results, source);

Comment (AkhtarAmir) [inline plugins/aws/ec2/outdatedAmi.js:27]: Check if (!config.check_ami_usage) return callback(null, results, source);

File: plugins/aws/ec2/outdatedAmi.js [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+            var describeLaunchConfigurations = helpers.addSource(cache, source,
+                ['autoscaling', 'describeLaunchConfigurations', region]);
+
+            if (!describeInstances || describeInstances.err || !describeInstances.data) {
+                helpers.addResult(results, 3,
+                    `Unable to query EC2 instances: ${helpers.addError(describeInstances)}`, region);
+                return rcb();
+            }
+
```
Comment (AkhtarAmir) [inline plugins/aws/ec2/outdatedAmi.js:66]: Add launch templates

File: plugins/aws/ec2/outdatedAmi.js [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+                helpers.addResult(results, 3,
+                    `Unable to query Auto Scaling launch configurations: ${helpers.addError(describeLaunchConfigurations)}`, region);
+                return rcb();
+            }
+            let found = false;
+            describeImages.data.forEach(image => {
+                const currentDate = new Date();
+                if (image.DeprecationTime && new Date(image.DeprecationTime) < currentDate) {                    
+                    found = true;
+                    if (config.check_ami_usage){
```
Comment (AkhtarAmir) [inline plugins/aws/ec2/outdatedAmi.js:78]: remove this check

File: plugins/aws/ec2/outdatedAmi.js [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+                    helpers.addResult(results, 2,
+                        `Amazon Machine Image "${image.ImageId}" is deprecated`,
+                        region, resource);
+                }
+            });
+
+            if (!found) {
+                helpers.addResult(results, 0,
+                    'No deprecated Amazon Machine Image found', region);
+            }
```
Comment (AkhtarAmir) [inline plugins/aws/ec2/outdatedAmi.js:118]: Remove this check

File: plugins/aws/ec2/outdatedAmi.js [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+                return rcb();
+            }
+            let found = false;
+            describeImages.data.forEach(image => {
+                const currentDate = new Date();
+                if (image.DeprecationTime && new Date(image.DeprecationTime) < currentDate) {                    
+                    found = true;
+                    if (config.check_ami_usage){
+                        let usedAmi = false;
+                        if (describeInstances.data.length) {
```
Comment (AkhtarAmir) [inline plugins/aws/ec2/outdatedAmi.js:80]: Change foreach to for loop and break from both loops if usedAmi is true

File: plugins/aws/ec2/outdatedAmi.js [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+                                    instance.Instances.forEach(element => {
+                                        if (element.ImageId && image.ImageId == element.ImageId) {
+                                            usedAmi = true;
+                                        }
+                                    });
+                                }
+                            });
+                        }
+            
+                        if (describeLaunchConfigurations.data.length) {
```
Comment (AkhtarAmir) [inline plugins/aws/ec2/outdatedAmi.js:92]:
```suggestion
                        if (!usedAmi && describeLaunchConfigurations.data.length) {
```

File: plugins/aws/ec2/outdatedAmi.js [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+                            describeLaunchConfigurations.data.forEach(config => {
+                                if (config.ImageId && image.ImageId == config.ImageId) {
+                                    usedAmi = true;
+                                }
+                            });
+                        }
+
```
Comment (AkhtarAmir) [inline plugins/aws/ec2/outdatedAmi.js:99]: 3 cases.
1. inuse + deprecated => fail
2. deprecated + notinuse => pass
3. not deprecated => pass

**Review by AkhtarAmir — CHANGES_REQUESTED** [review]

**Review by AkhtarAmir — APPROVED** [review]

File: plugins/aws/ec2/outdatedAmiInUse.js [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+                if (image.DeprecationTime && new Date(image.DeprecationTime) < new Date()) {                    
+                    found = true;
+                    let amiInUse = false;
+
+                    if (describeInstances.data.length) {
+                        for (const instance of describeInstances.data) {
+                            if (instance.Instances && instance.Instances.length) {
+                                for (const element of instance.Instances) {
+                                    if (element.ImageId && image.ImageId == element.ImageId) amiInUse = true;
```
Comment (giorod3) [inline plugins/aws/ec2/outdatedAmiInUse.js:81]: break should be inside the if statement

File: plugins/aws/ec2/outdatedAmiInUse.js [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+                                    break;
+                                }
+                            }
+                            if (amiInUse) break;
+                        }
+                    }
+        
+                    if (!amiInUse && describeLaunchConfigurations.data.length) {
+                        for (const config of describeLaunchConfigurations.data) {
+                            if (config.ImageId && image.ImageId == config.ImageId) amiInUse = true;
```
Comment (giorod3) [inline plugins/aws/ec2/outdatedAmiInUse.js:91]: break should be inside the if statement

**Review by giorod3 — CHANGES_REQUESTED** [review]
just the couple of comments

**Review by giorod3 — APPROVED** [review]

**Pull request merged**